src/Microsoft.ML.TorchSharp/Roberta/Models/RobertaModel.cs

```diff
@@ -0,0 +1,133 @@
+            Ones = Enumerable.Repeat(1, options.MaxSequenceLength).ToArray();
+            NegBillionPad = Enumerable.Repeat(negBillion, options.MaxSequenceLength).ToArray();
+
+            Encoder = new RobertaEncoder(
+                numLayers: 12,
+                numAttentionHeads: 12,
+                numEmbeddings: 50265,
+                embeddingSize: 768,
+                hiddenSize: 768,
+                outputSize: 768,
+                ffnHiddenSize: 3072,
+                maxPositions: 512,
+                maxTokenTypes: 2,
+                layerNormEps: 1e-12,
+                embeddingDropoutRate: 0.1,
+                attentionDropoutRate: 0.1,
+                attentionOutputDropoutRate: 0.1,
+                outputDropoutRate: 0.1);
+        }
+
+        protected void InitWeights(torch.nn.Module module)
+        {
+            using var disposeScope = torch.NewDisposeScope();
+            if (module is Linear linearModule)
+            {
+                linearModule.weight.normal_(mean: 0.0, std: 0.02);
+                if (linearModule.bias.IsNotNull())
+                {
+                    linearModule.bias.zero_();
+                }
+            }
+            else if (module is Embedding embeddingModule)
+            {
+                embeddingModule.weight.normal_(mean: 0.0, std: 0.02);
+                embeddingModule.weight[1].zero_();  // padding_idx
+            }
+            else if (module is LayerNorm layerNormModule)
+            {
+                layerNormModule.weight.fill_(1.0);
+                layerNormModule.bias.zero_();
+            }
+        }
+
+        /// <summary>
+        /// Run only Encoder and return features.
+        /// </summary>
+        protected torch.Tensor ExtractFeatures(torch.Tensor srcTokens)
+        {
+            var (positions, segments, attentions) = GetEmbeddings(srcTokens);
+            var encodedVector = Encoder.call(srcTokens, positions, segments, attentions);
+            return encodedVector;
+        }
+
+        private (torch.Tensor position, torch.Tensor segment, torch.Tensor attentionMask) GetEmbeddings(torch.Tensor srcTokens)
+        {
+            using var disposeScope = torch.NewDisposeScope();
+            var device = srcTokens.device;
+            var positions = new torch.Tensor[srcTokens.size(0)];
+            var segments = new torch.Tensor[srcTokens.size(0)];
+            var attentionMasks = new torch.Tensor[srcTokens.size(0)];
+
+            for (var i = 0; i < srcTokens.size(0); ++i)
+            {
+                var srcTokenArray = srcTokens[i].ToArray<int>();
+                var size = srcTokenArray.Length;
+                var questionSize = srcTokenArray.ToList().IndexOf(EosIndex) - 1;
+                var allSize = srcTokenArray.Count(token => token != PadIndex);
+                var position = torch.tensor(Range.GetSubArray(Positions, 0..allSize).Concat(Range.GetSubArray(Zeros, ..(size - allSize))).ToArray(),
```

Comment:
This looks like a lot of allocating and copying.  Seems like ideally we'd just do one allocation as torch.tensor and copy from existing managed memory.  That might alleviate the need for GetSubArray entirely.  cc @tarekgh

Reply:
Resolved my other feedback -- this looks like the place to focus your efforts.  Lots of extra allocation here.  Instead lets see if we can create a single array based on composing spans.

Reply:
Have switched them to using span/span concat instead.